### PR TITLE
CFY-3009 local workflow test decorator

### PR DESCRIPTION
**Pull request by max-orlov**
CFY-3009 local workflow test decorator. enables to set different values for the local workflow - only blueprint and source file path are mandatory.

### CODE REVIEW

File: cloudify/tests_resources/__init__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+########
+# Copyright (c) 2014 GigaSpaces Technologies Ltd. All rights reserved
```
Comment (idanmo) [inline cloudify/tests_resources/__init__.py]: 2015

File: cloudify/tests_resources/local_workflows_decorator.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
+########
+# Copyright (c) 2014 GigaSpaces Technologies Ltd. All rights reserved
```
Comment (idanmo) [inline cloudify/tests_resources/local_workflows_decorator.py]: 2015

File: cloudify/tests_resources/local_workflows_decorator.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+# * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# * See the License for the specific language governing permissions and
+# * limitations under the License.
+
+import shutil
+import tempfile
+from os import path, listdir
+
+from cloudify.workflows import local
+from functools import wraps
```
Comment (idanmo) [inline cloudify/tests_resources/local_workflows_decorator.py]: separate system imports with 3rd party imports, in other words functools should place in the first chunk

File: cloudify/tests_resources/local_workflows_decorator.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def _find_plugin_yaml(file_path):
+    if 'plugin.yaml' in listdir(file_path):
```
Comment (idanmo) [inline cloudify/tests_resources/local_workflows_decorator.py]: store plugin.yaml as a module constant...

File: cloudify/tests_resources/local_workflows_decorator.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        return path.join(file_path, 'plugin.yaml')
+    return _find_plugin_yaml(path.realpath(path.join(file_path, '..')))
```
Comment (idanmo) [inline cloudify/tests_resources/local_workflows_decorator.py]: what happens if plugin.yaml is not present at all in the folder tree?
doesn't seem like you're handling that case...

File: cloudify/test_utils/local_workflow_decorator.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        if level_up_path == running_path:
+            msg = 'Traversing up the folder tree from {0}, failed to find {1}.'
+            raise IOError(msg.format(original_path, PLUGIN_NAME))
+        else:
+            running_path = level_up_path
+
+    return path.join(running_path, PLUGIN_NAME)
+
+
+class set_testing_env(object):
```
Comment (max-orlov) [inline cloudify/test_utils/local_workflow_decorator.py]: change to workflow_test

File: cloudify/test_utils/local_workflow_decorator.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+#  * limitations under the License.
+
+import shutil
+import tempfile
+from os import path, listdir
+from functools import wraps
+
+from cloudify.workflows import local
+
+PLUGIN_NAME = 'plugin.yaml'
```
Comment (max-orlov) [inline cloudify/test_utils/local_workflow_decorator.py]: change name

File: cloudify/test_utils/local_workflow_decorator.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+            raise IOError(msg.format(original_path, PLUGIN_NAME))
+        else:
+            running_path = level_up_path
+
+    return path.join(running_path, PLUGIN_NAME)
+
+
+class set_testing_env(object):
+    def __init__(self,
+                 source_file_path,
```
Comment (max-orlov) [inline cloudify/test_utils/local_workflow_decorator.py]: move to set_up

File: cloudify/test_utils/local_workflow_decorator.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                 blueprint_relative_path,
+                 copy_plugin=False,
```
Comment (max-orlov) [inline cloudify/test_utils/local_workflow_decorator.py]: copy_plugin_yaml

File: cloudify/test_utils/local_workflow_decorator.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+#  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  * See the License for the specific language governing permissions and
+#  * limitations under the License.
+
+import sys
+import shutil
+import tempfile
+from os import path, listdir, makedirs
+
+from functools import wraps
```
Comment (max-orlov) [inline cloudify/test_utils/local_workflow_decorator.py]: push up

File: cloudify/test_utils/local_workflow_decorator.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+            resource_source_path = resource
+            relative_dest_path = path.basename(resource_source_path)
+
+        resource_source_path = path.abspath(resource_source_path)
+        resource_dest_path = path.join(test_root_path, relative_dest_path)
+        _assure_path_exists(path.dirname(resource_dest_path))
+        shutil.copyfile(resource_source_path, resource_dest_path)
+
+
+# TODO: env_as_keyword not tested yet (as the whole project is in python2 only)
```
Comment (max-orlov) [inline cloudify/test_utils/local_workflow_decorator.py]: no need

File: cloudify/test_utils/local_workflow_decorator.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+        :param local_file_path: the path of the test file.
+        :param test_method_name: the name of the test method.
+        :return: The test env which is a wrapped Environment.
+        """
+        if not self.prefix:
+            self.prefix = test_method_name
+
+        # Creating temp dir
+        self.temp_dir = tempfile.mkdtemp(self.prefix)
```
Comment (max-orlov) [inline cloudify/test_utils/local_workflow_decorator.py]: prefix=

File: cloudify/test_utils/local_workflow_decorator.py [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+            The wrapper function itself.
+
+            :param func: the function of which this test has been called from
+            :param args:
+            :param kwargs:
+            :return:
+            """
+            test_env = self.set_up(
+                sys.modules[func.__class__.__module__].__file__,
+                test.__name__
```
Comment (max-orlov) [inline cloudify/test_utils/local_workflow_decorator.py]: test._testMethodName

File: cloudify/test_utils/local_workflow_decorator.py [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+        # Finding and adding the plugin
+        if self.copy_plugin:
+            self.resources_to_copy.append(
+                _find_plugin_yaml(path.dirname(local_file_path)))
+
+        # Copying resources
+        _copy_resources(self.temp_dir, self.resources_to_copy)
+
+        # Updating the test_method_name (if not manually set)
+        if self.init_args and not self.init_args.get('name', False):
```
Comment (max-orlov) [inline cloudify/test_utils/local_workflow_decorator.py]: without explicit default value

File: cloudify/test_utils/local_workflow_decorator.py [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+    """
+    Tries to find the plugin.yaml file automatically (by traversing up the
+    directory tree).
+    :param original_path: The path to start the search from
+    :return: The absolute path of the plugin.yaml file (if found, o.w.
+    raises an Error)
+    """
+    running_path = original_path
+    while PLUGIN_YAML_NAME not in listdir(running_path):
+        level_up_path = path.realpath(path.join(running_path, '..'))
```
Comment (dankilman) [inline cloudify/test_utils/local_workflow_decorator.py]: os.path.dirname

**Pull request merged**